WIP: The Avalonia port

luizzeroxis opened the pull request:
[Downloads for macOS, Linux and Windows linked here](https://nightly.link/luizzeroxis/UndertaleModTool/workflows/publish_gui_avalonia/avalonia)

---

**STATUS:** A few features are missing.

---

I've been working on the Avalonia port of UndertaleModTool for a little while, so I decided to finally make a draft PR for it, in case anyone else wants to help. As discussed in #460 and elsewhere, Avalonia is probably the best choice, since it's multi-platform, works in C# and seems to suck less than alternatives. I tested the current build on Linux and it seems to work fine (except the performance).

Currently, I'm working on making it work as a sort of 'view only' UTMT, so those not on Windows can datamine games, at least. Of course, the architecture still has to be made such that we could later add editing functionality.

But, it is not complete yet. No, it is far from complete.  Here's a TODO list of all the things I can think of that need to be done (but there's certainly more):

- General
  - [x] Open data file
  - [x] Save data file
    - [ ] Save debug information
  - [ ] Temp run
  - [x] Middle click to open in new tab
  - [x] Keyboard shortcuts
  - [ ] Make sure null and non-existent items work

- Main UI
  - [x] Tree view
    - [x] Filter by name
    - [x] Virtualization
    - [x] Right click menu
    - [x] Reorder resources
  - [x] Tabs
    - [x] Back/forward navigation
    - [x] Open in the same tab
    - [x] Keep tab state
  - [x] Resource ID number

- Editors
  - [x] General info editor
  - [x] Global init scripts editor
  - [x] Game end scripts editor
  - [x] Audio group editor
  - [x] Sound editor
    - [x] Play audio
    - [x] Deal with audio groups
  - [x] Sprite editor
    - [x] Show first image initially
  - [x] Background editor
  - [x] Path editor
    - [x] Render path
  - [x] Script editor
  - [x] Shader editor
  - [x] Font editor
    - [ ] Glyph editor
  - [x] Timeline editor
  - [x] Game object editor
  - [x] Room editor
    - [x] Background properties
    - [x] View properties
    - [x] Instance properties
    - [x] Tile properties
    - [x] Layer properties
    - [x] Render all items
    - [ ] Edit all items
    - [x] Select and move items
    - [ ] Scale and rotate items
    - [x] Tiles layer editing
    - [x] GMS 1 tile editing
    - [x] Save to image
    - [x] Reorder items
    - [x] Copy and paste items
    - [x] Gray out disabled items
  - [x] Extension editor
    - [x] Product ID
  - [x] Texture page item editor
    - [ ] Show item in texture
  - [x] Code editor
    - [x] Syntax highlighting
    - [x] Parent entry
    - [x] Right click stuff
  - [x] Variable editor
  - [x] Function editor
  - [x] Code local editor
  - [x] String editor
  - [x] Embedded texture editor
    - [ ] Click item to open
  - [x] Embedded audio editor
  - [x] Texture group information editor
  - [x] Embedded image editor
  - [x] Particle system editor
  - [x] Particle system emitter editor

- Controls
  - [x] String reference box
    - [ ] Choose to replace one/all
  - [x] Resource reference box
  - [x] Flags box
    - [x] Text box
  - [x] Editable data grid
  - [x] Color picker
  - [x] Image viewer

- Other
  - [x] Basic message dialogs
  - [x] Text input dialog
  - [x] Loader dialog
  - [x] Settings
    - [ ] All settings
  - [x] Search in code
  - [x] Find references
  - [ ] File associations
  - [x] Icon

- Scripting
  - [x] Be able run scripts
  - [x] Command textbox
  - [ ] Full scripting API

- Project
  - [x] Open/save projects
  - [x] View unexported assets window
  - [x] View and edit code entry GML
  - [ ] Auto mark for export when editing resources

luizzeroxis committed:
Forgot to add this file

luizzeroxis committed:
Add collision mask viewer

luizzeroxis committed:
Add settings window

luizzeroxis committed:
Add particle system editor

luizzeroxis committed:
Add particle system emitter editor

luizzeroxis committed:
Auto select first item on data grid

luizzeroxis committed:
Add extension editor

HappyRespawnanchor commented:
nice work :)

luizzeroxis committed:
Merge remote-tracking branch 'origin/master' into avalonia

luizzeroxis committed:
Added (rough) search in code window

luizzeroxis committed:
Open search result in line, enter to search (shift+enter to add new line), enter to open result, highlight current line in code editor

luizzeroxis committed:
Open file through command line argument, show data file errors

luizzeroxis committed:
Change Version to DataVersion

luizzeroxis committed:
Save settings to file and set version

luizzeroxis committed:
Update dependencies

luizzeroxis committed:
Basic scripting support, add command box funcionality

luizzeroxis committed:
Import and export collision mask data

luizzeroxis committed:
Added more scripting functions

luizzeroxis committed:
Small accumulated fixes

luizzeroxis committed:
Change bug fix to setting AutoScrollToSelectedItem to false since that works better

luizzeroxis committed:
Added tab navigation, open in new tab options, changed some button text for compactness, focus on textbox in search in code

luizzeroxis committed:
Add draggable tabs

luizzeroxis committed:
Added loader dialog

luizzeroxis committed:
Add more scripting API implementations and improve scripting

luizzeroxis committed:
Add SkiaSharp.NativeAssets.Linux

luizzeroxis committed:
Add remove and copy name options to resources

luizzeroxis committed:
Room editor improvements

luizzeroxis committed:
Various improvements
Copy message dialogs with Ctrl+C, search in code shows errors in status bar, opening results no longer changes window, context menu font sizes normalized, small layout spacing changes, add close data file command, tab right click

luizzeroxis committed:
Add room editors for sprite and particle system instances

luizzeroxis commented:
@MayaChen350 Hi, you're welcome to send PRs to my repo for any features that are missing, thanks

MayaChen350 commented:
I'll do my best then! I switched to Fedora recently but then it happened I still wanted to make mods for Deltarune, so this PR could be a life saver

Thank you

luizzeroxis committed:
Add icon

luizzeroxis committed:
Room editor fixes
Fix crashes for zero-sized tiles, fix scaled tiles, fix wrongly scaled instances with non-zero origin, sort layers and tiles by depth, add position property to tiles

luizzeroxis committed:
Click on room items to select them

luizzeroxis committed:
Add loader dialogs on code editor

luizzeroxis committed:
Select sprite instances in room, reset translation and scaling when changing rooms, press F to scroll to selected item in room, fix title not updating

luizzeroxis committed:
Avalonia GitHub Actions

luizzeroxis committed:
Move items in room

luizzeroxis committed:
Add decompile settings

luizzeroxis committed:
Merge remote-tracking branch 'origin/master' into avalonia

luizzeroxis committed:
Only select from selected category

luizzeroxis committed:
Properly select selected item in room tree, keep current layer when unselecting

luizzeroxis committed:
Ask user for asset name when creating, and if room should be added in room order

luizzeroxis committed:
Add scripting function for text inputs

luizzeroxis committed:
Flip and rotate tiles, instance color, fix crash with rooms after 2024.13

luizzeroxis committed:
Fix data grid add not enabling

luizzeroxis committed:
Save changes before quitting dialog

luizzeroxis committed:
Small dialog fixes

luizzeroxis committed:
Ask user to save when creating new, opening or closing data file

luizzeroxis committed:
Moved view accessing functions to interface

luizzeroxis committed:
Loader window is now an interface for testing

luizzeroxis committed:
Merge remote-tracking branch 'origin/master' into avalonia

luizzeroxis committed:
Update version to 0.8.3.0

luizzeroxis committed:
Allow custom Styles.xaml in AppData

luizzeroxis committed:
Moved room properties to left side, zoom textbox

luizzeroxis committed:
Fix forgotten Avalonia.Markup.Xaml.Loader and update dependencies

luizzeroxis committed:
Single namespace

luizzeroxis committed:
Reorganized folders

luizzeroxis committed:
Restore caret position and focus after compile in code editor

luizzeroxis committed:
Basic code syntax highlighting

luizzeroxis committed:
Rough tile editor

luizzeroxis committed:
Add copy and paste to room instances

luizzeroxis committed:
Add move up and down buttons to room editor

luizzeroxis committed:
Add swap to position to resources

luizzeroxis committed:
Removed room debug text, added room coordinates under mouse

luizzeroxis committed:
Setting to disable syntax highlighting

luizzeroxis committed:
Update dependencies

luizzeroxis committed:
Merge branch 'master' into avalonia

luizzeroxis committed:
Fix for updates

luizzeroxis committed:
Add option to not auto compile

luizzeroxis committed:
Fix sprites not showing up when opening

luizzeroxis committed:
Fix color background

luizzeroxis committed:
Fix negative scaled items not being selected in room

luizzeroxis committed:
Render tiled backgrounds

luizzeroxis committed:
Room editor reorganize around room items

luizzeroxis committed:
Changed custom data tree view with TreeDataGrid
This definitely won't cause any problems in the future!

luizzeroxis committed:
Temporary fix to resource position swaps sometimes not working

luizzeroxis committed:
Setting to open new resource after creating it

luizzeroxis committed:
Fix crash when rendering happens at same time as updates in room editor

luizzeroxis committed:
Add room grid

luizzeroxis committed:
Removed unecessary columns and buttons in data grids

luizzeroxis committed:
Remove DynamicData, replace with custom solution

luizzeroxis committed:
Update dependencies

luizzeroxis committed:
Added a reset to the list when there's too many changes to fix performance.

luizzeroxis committed:
Gray parented code entries and null entries (again)

luizzeroxis committed:
Update SKImageViewer to be simpler and more robust

luizzeroxis committed:
Show sprite in object, hide GMS2 stuff in GMS1.

luizzeroxis committed:
Merge branch 'master' into avalonia

luizzeroxis committed:
Update to .NET 10

luizzeroxis committed:
Add right click to code editor to replace integer index with a proper name, and to open resources (it's still a bit messy). Fix highlighting in old versions that lack string escaping.

luizzeroxis committed:
Improved code editor performance by caching highlighting data

luizzeroxis committed:
Added loader to search on code

luizzeroxis committed:
Expand data tree item on load

luizzeroxis committed:
View font and embedded image image

luizzeroxis committed:
Import audio

luizzeroxis committed:
Import embedded texture

luizzeroxis committed:
Drop data files into window to open

luizzeroxis committed:
CollectionChanged and loader window works in other threads (hopefully improves scritpt compatibility)

luizzeroxis committed:
Removed ShowMessageDialog

luizzeroxis committed:
Add right click option on tab to select item on list

luizzeroxis committed:
Moved ScrollViewer inside views to fix some issues

luizzeroxis committed:
Add save as image to room editor

luizzeroxis committed:
Add button to export sprite textures as PNGs

luizzeroxis committed:
Fast tile rendering with DrawVertices

luizzeroxis committed:
Fix whitespace

luizzeroxis committed:
Merge branch 'master' into avalonia

luizzeroxis committed:
Update for new background properties

luizzeroxis committed:
Add path viewer

luizzeroxis committed:
Little reorganizing of room editor

luizzeroxis committed:
Add select any layer to room editor

luizzeroxis committed:
Merge remote-tracking branch 'origin/master' into avalonia

luizzeroxis committed:
Update version number

luizzeroxis committed:
Update dependencies

luizzeroxis committed:
Add tile picker

luizzeroxis committed:
Formatting fixes

luizzeroxis committed:
Move room view things to view model

luizzeroxis committed:
Add auto size tile layer button

luizzeroxis committed:
Use resources in styles

luizzeroxis committed:
Tile set columns control now a NumericUpDown

luizzeroxis committed:
Add some keyboard shortcuts

luizzeroxis committed:
Add middle click button in places

luizzeroxis committed:
Add flip x, flip y and rotate buttons to tile picker, and right click to delete tile

luizzeroxis committed:
Room editor reorganizing

luizzeroxis committed:
Fix crash after changing syntax highlight setting (#7)

luizzeroxis committed:
Moved room renderer to separate file

luizzeroxis committed:
Limiting room items and isolating custom draw operation

luizzeroxis committed:
The room renderer now first creates render commands, which will store the state of the room before reading them in the render thread. This is for thread safety, so that it'll be consistent even if the data changes while it is rendering.

luizzeroxis committed:
Cull tiles in tiles layer outside view

luizzeroxis committed:
Changed to software rendering on Linux to hopefully avoid crashes

luizzeroxis committed:
Add animation curve editor

luizzeroxis committed:
Only show chunks that exist in data file

luizzeroxis committed:
Made selection color the same as theme in custom draw operations

luizzeroxis committed:
Small changes

luizzeroxis committed:
Add textbox to flags box

luizzeroxis committed:
Add "run with runner" feature

luizzeroxis committed:
Change MessageWindow's arguments to Buttons

luizzeroxis committed:
Moved file picker types to separate file

luizzeroxis committed:
Forgot other operating systems existed

luizzeroxis committed:
Added basic test project

skirlez committed:
Make double tapping reference views go to the reference

luizzeroxis committed:
Merge branch 'master' into avalonia

luizzeroxis committed:
Add audio player with SDL

luizzeroxis committed:
Add attach and detach events to tabs, sounds stop when embedded audio tab is detached.

luizzeroxis committed:
View and play audio from sound if in the same data file

luizzeroxis committed:
Fix version comparer

luizzeroxis committed:
Improve audio player

luizzeroxis committed:
Improve some rotations

luizzeroxis committed:
Zoom in tile picker

luizzeroxis committed:
Checkered background in tile picker

luizzeroxis committed:
Try to fix crash in large rooms

luizzeroxis committed:
Use default monospace font on Linux, moved font config to style resource

luizzeroxis committed:
Quick fix for search results not opening in correct line when compilation takes too long

luizzeroxis committed:
Slightly better go to location logic

luizzeroxis committed:
Add add button to reference, can add new code entry within object with normal name.

BioTomateDE commented:
> (for macos)

mac sucks, u either gotta compile it yourself (easier) or somehow allow unsigned applications lol

luizzeroxis committed:
Add setting to not start maximized

luizzeroxis committed:
Try to show message on error

luizzeroxis committed:
Moved things to ImportExport

luizzeroxis committed:
Fix and add tests to ObservableCollectionView

luizzeroxis committed:
Binary search in observable collection view, more testing

luizzeroxis committed:
Allow shift+enter in command box

luizzeroxis committed:
Add some TargetWidth and TargetHeight support

luizzeroxis committed:
Merge branch 'master' into avalonia

luizzeroxis committed:
Update dependencies

luizzeroxis committed:
Allow playing sounds in sound editor if there's no audio groups

luizzeroxis committed:
Merge branch 'skirlez/avalonia' into avalonia

luizzeroxis committed:
Allow dragging and dropping objects and sprites into room

luizzeroxis committed:
Add auto tile IDs button

luizzeroxis committed:
Small scripting fixes

luizzeroxis committed:
Merge commit 'f3f1c878' into avalonia

luizzeroxis committed:
Fix things from last merge

luizzeroxis committed:
Merge commit '8cae55567019123d3d1f33cd7aebd2d3e6dfc4c0' into avalonia-project-system

luizzeroxis committed:
Fix errors caused by merge

luizzeroxis committed:
Merge branch 'master' into avalonia-project-system

5GameMaker commented:
BUG: After opening a new file, search is not applied until it's modified

luizzeroxis committed:
Add basic project file management. New, open, save and close projects.

luizzeroxis committed:
Add marked for export checkbox, mark add and unmark removed, fix resource number not changing

luizzeroxis committed:
Fix project data file saving

luizzeroxis committed:
Source code from project can be viewed and edit in code editor and be searched.

CST1229 commented:
random thing i found: middle-clicking or ctrl+clicking numbers in code doesnt seem to work in this version (unlike in the current main version), only rightclicking
(similar for functions/assets, where only middle click works and not ctrl+click)

luizzeroxis committed:
Merge branch 'master' into avalonia-project-system

luizzeroxis committed:
Update version to 0.9.0.0

luizzeroxis committed:
Updated dependencies, moved to wieslawsoltes's TreeDataGrid for Avalonia 12 support

luizzeroxis committed:
Control + left click to open code resource in new tab

luizzeroxis committed:
Update filter when changing data file

luizzeroxis committed:
Small formatting fixes

CST1229 commented:
also, ctrl+s always shows the save dialog at the home directory with nothing selected (instead of immediately saving the data file to its last location if one was loaded), at least for me

ephoton commented:
On macOS, keyboard shortcuts should use `Cmd` instead of `Ctrl`

sjl057 commented:
i'm getting really bad lag when trying to display a lot of text in scripts. for what it's worth, i have a high dpi monitor

luizzeroxis committed:
Trying to fix services

luizzeroxis committed:
MacOS shortcuts

luizzeroxis committed:
Add project to title bar

luizzeroxis committed:
Increase max GPU resource size to 512 MB

luizzeroxis committed:
Workaround for slowdown in code editor on Linux

devek1 commented:
I got an instant error on trying to launch the app from KDE Dolphin file manager:
`The desktop entry file /home/devek1/GamingTools/UTMT Native/UndertaleModToolAvalonia.Desktop has no Type=... entry.`

In order to fix this, the file has to simply be renamed with an appropriate extension, such as .x86-64 (or alternatively, removing the extension entirely), so that it is correctly opened as a binary rather than a desktop file.

sjl057 commented:
The package on the AUR also doesn't pass the validity check

On Fri, Jun 5, 2026 at 11:55 AM devek1 ***@***.***> wrote:

> *devek1* left a comment (UnderminersTeam/UndertaleModTool#2126)
> <https://github.com/UnderminersTeam/UndertaleModTool/pull/2126#issuecomment-4633687340>
>
> I got an instant error on trying to launch the app from KDE Dolphin file
> manager:
> The desktop entry file /home/devek1/GamingTools/UTMT
> Native/UndertaleModToolAvalonia.Desktop has no Type=... entry.
>
> In order to fix this, the file has to simply be renamed with an
> appropriate extension, such as .x86-64 (or alternatively, removing the
> extension entirely), so that it is correctly opened as a binary rather than
> a desktop file.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/UnderminersTeam/UndertaleModTool/pull/2126?email_source=notifications&email_token=AB6WUF6GET5O7Z44PMFF2JT46L3RVA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTINRTGM3DQNZTGQYKM4TFMFZW63VHMNXW23LFNZ2KKZLWMVXHJLDGN5XXIZLSL5RWY2LDNM#issuecomment-4633687340>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB6WUFYYFSDSHXM4BUMSD3D46L3RVAVCNFSM6AAAAAB5L5ISASVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHM2DMMZTGY4DOMZUGA>
> .
> Triage notifications, keep track of coding agent tasks and review pull
> requests on the go with GitHub Mobile for iOS
> <https://github.com/notifications/mobile/ios/AB6WUF2G5P6F6UYFYA4BYRD46L3RVA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTINRTGM3DQNZTGQYKM4TFMFZW63VHMNXW23LFNZ2KKZLWMVXHJKTGN5XXIZLSL5UW64Y>
> and Android
> <https://github.com/notifications/mobile/android/AB6WUFZ22NOUWNN57X6ZIJT46L3RVA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTINRTGM3DQNZTGQYKM4TFMFZW63VHMNXW23LFNZ2KKZLWMVXHJLTGN5XXIZLSL5QW4ZDSN5UWI>.
> Download it today!
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

Damglador commented:
AUR comment section would be a better place to report that. The package shouldn't have a validity check in the first place, so I've already sent a patch to fix that there.

luizzeroxis committed:
Add unexported assets window

luizzeroxis committed:
Combined Desktop project into main project

luizzeroxis committed:
Merge branch 'master' into avalonia

luizzeroxis committed:
Updated to 0.9.1.0 and updated dependencies

luizzeroxis committed:
Add lock button to room editor

luizzeroxis committed:
Small fixes

luizzeroxis committed:
Add bad find references window

luizzeroxis committed:
Add context menu option to find references

luizzeroxis committed:
Add loader window to find references

luizzeroxis committed:
Add version to places

luizzeroxis committed:
Add message to code editor warning if code is compiled or not

luizzeroxis committed:
Improved ObservableCollectionView code

luizzeroxis committed:
Attempt to fix focus and UI disabling problems

luizzeroxis committed:
Remove image pinning

luizzeroxis committed:
Go to column in search in code

luizzeroxis committed:
Remember last data location

luizzeroxis committed:
Add external texture controls to texture group info view

luizzeroxis committed:
Add ctrl+scroll to scale images

luizzeroxis committed:
Add some kind of nine patch support to rendering

luizzeroxis committed:
Add support to playing sounds in audio group files

luizzeroxis committed:
Add git version to about box

luizzeroxis committed:
Add run with data/project runner menu item (F5)

luizzeroxis committed:
Merge branch 'master' into avalonia

luizzeroxis committed:
Update to 0.9.1.1 and cleanups

luizzeroxis committed:
Updated dependencies

luizzeroxis committed:
Gray out disabled items in room editor list

luizzeroxis committed:
Separate data explorer view

luizzeroxis committed:
Add add button to string references and some tooltips.

luizzeroxis committed:
Added a button to sort data list alphabetically (had to change the entire ObservableCollectionView for this)

luizzeroxis committed:
Add filter delay

luizzeroxis committed:
Moved things to data explorer view model

luizzeroxis committed:
Merge branch 'master' into avalonia

luizzeroxis committed:
Update to 0.9.1.2 and update dependencies

luizzeroxis committed:
Fix some bugs with ObservableCollectionView

luizzeroxis committed:
Fix crash when going out of bounds in room with tiles

luizzeroxis committed:
Removed AllAssets from UndertaleRoom

luizzeroxis committed:
Try to fix double selection when using select on list on tab, it still looks wrong though...

luizzeroxis committed:
Add toggle visiblity menu item to layers in rooms

luizzeroxis committed:
Move from canton7's PropertyChanged.SourceGenerator to MVVM Toolkit

luizzeroxis committed:
Fix some problems with nine slices

luizzeroxis committed:
Add hotkeys to tile editing

luizzeroxis committed:
Add setting to disable backups

luizzeroxis committed:
Don't set code to outdated if it just decompiled or loaded

luizzeroxis committed:
Add nine slice fields to sprite

luizzeroxis committed:
Use compiled bindings on item template, use dynamic only at the last possible moment in flags box, fix useless warning about the text box window

luizzeroxis committed:
Improved code editor (and removed undo)

luizzeroxis committed:
Improvements for code view and search in code go to location

luizzeroxis committed:
Tab save system - when closing the tab or saving the data, it'll try to finish compiling code entries before continuing

luizzeroxis committed:
Add extension chunk editor

luizzeroxis committed:
Set highlight code editor color

luizzeroxis committed:
Fix not being able to set string references

luizzeroxis committed:
Add several checks for null resource names

luizzeroxis committed:
Add .app bundle build

luizzeroxis committed:
Add app name and remove app menu

luizzeroxis committed:
Add tile picker for GMS1 tiles, make selected border in room keep same thickness when zoomed, add meters per pixel,

luizzeroxis committed:
Fix duplicating flags

luizzeroxis committed:
Alt drag to select region in room tile picker

luizzeroxis committed:
Shift to override grid enabled/disabled

luizzeroxis committed:
Add tile editing to legacy tiles

luizzeroxis committed:
Changed string reference behavior, now it only changes the current field, creating a new string if necessary. Ctrl+enter updates it

luizzeroxis committed:
Dark window decorations on Windows 10

luizzeroxis committed:
Deleted tests

luizzeroxis committed:
Add padding to load/save in texture page item

luizzeroxis committed:
Try to fix issues with loading project code

luizzeroxis committed:
Add reload project menu item
